Standard module options

reidmv opened the pull request:
This PR is primarily designed to sanitize the schema used by the Hash source type to define environments and modules, adopting a standard way of describing content (environments, modules) which is not implementation-dependent, the way today's modules especially are.

The idea is that when specifying content in a data hash, it should be possible to specify it using these standard keys.

```yaml
---
object-name:
  type: <string> # E.g. "git", "forge", "svn"
  source: <string> # E.g. "git://server.tld/project/repo.git", "https://forge.puppet.com", etc.
  version: <string> # E.g. "main", "06a32bc", "6.1.0", etc.
```

This PR assumes #1130 exists to improve the visibility of what happens if users use conflicting ways of specifying values like version, such as by using both `version` and an implementation-specific key like `git`, but does not depend on that improvement.

This PR homogenizes how options are parsed for many objects by standardizing on the use of `R10K::Util::Setopts`.

See the [rich diff of the doc/dynamic-environments/configuration.mkd](https://github.com/puppetlabs/r10k/pull/1131/files?short_path=5a41a0b#diff-5a41a0b389c98bce0313a7b6e74d1dcb2a7dfb8e09d5c7f983c33b9fb46c9ec6) page for the best visual of what this change accomplishes for consumers.

This improvement takes a UX step in the direction of r10k being able to support other types of environment sources and module sources, such as e.g. tarballs from Artifactory or S3.

This PR should be backwards compatible, without behavior changes, with one exception. Some object types did not validate schema in any way for keys passed into them. `R10K::Util::Setopts` does validate schema, and so garbage keys that would have been ignored previously may now cause an error.

reidmv committed:
Implement standard module option schema

Allow modules of any currently supported type to be initialized using
the following standard specification:

  ---
  <module-name>:
    type: <type>
    source: <source>
    version: <version>

This is in slight contrast to today's expectations, which are largely
based on the presence of a type-specific key to indicate the kind of
module to create, and there are no standard parameters to indicate
source and version.

By providing a standard schema for defining modules of any type, this
paves the way for easier integration with external data systems which
may define sets of modules of many types to include in different
environments.

This schema is generalized away from implementation and towards the
concepts that matter to the user.

reidmv committed:
Standardize environment option interface

And make it obvious which options are part of the standard interface.

Clean up :puppetfile_name handling

The pattern appears to be, call super first. The base class initialize
method saves off the original, unmodified options to @args. Base classes
then delete options they process. The outer class uses setopts to
process the remainder.

reidmv committed:
Document new general options for hash sources

Add standard interface updates to docs, changelog

Co-authored-by: Molly Waggett <mwaggett@alumni.reed.edu>

justinstoller commented on lib/r10k/module/svn.rb:
```diff
@@ -47,7 +52,7 @@ def self.implement?(name, args)
     super
 
-    setopts(opts, INITIALIZE_OPTS)
+    setopts(@args, INITIALIZE_OPTS)
```
IMO, stylistically it's better to just use the parameter name you're passed into, even it is effectively the same as an instance variable created elsewhere. I don't know if that's actually

reidmv replied:
Yeah. The main reason this goes to `@args` is the sanitization performed on the instance variable [here](https://github.com/puppetlabs/r10k/pull/1131/files#diff-db1c390dda7d3b2f14abcee49a787a9a31ffa9a41d2a0cbea53e35fb65b86773R55), to remove a common key "type" so that individual `setopts()` calls don't have to care if it's present.

One alternative option could be to provide `setopts()` with the ability to ignore certain keys. If we did that, we'd have to ignore it everywhere, but also wouldn't have to sanitize `@args`. E.g.,

```
setopts(options, {
  # Standard option interface
  :source  => :remote,
  :version => :expected_revision,
  :type => :ignore

  # Type-specific options
  :remote => :self,
  :username => :self,
  :password => :self,
  :puppetfile_name => :self,
})
```

Think that would be better?

reidmv replied:
Pushed a new commit with this change. Seems more straightforward, and therefore better.

mwaggett commented on doc/puppetfile.mkd:
```diff
@@ -109,11 +109,20 @@ latest version available.
+      type:    'forge',
+      version: '6.0.0'
+
+Note that at present the `:source` parameter is permitted, but not implemented.
```
I'm not sure what you mean by this.

reidmv replied:
I think based on that, I'm going to remove it, and just say that a `:source` parameter isn't supported for individual forge modules. At present, r10k will only let you specify a Forge source once, globally. We could add individual forge sources, but there isn't a use case reason to right now.

Edit: oh wait, I remember now why it was there. Because the whole point of this update is to make it possible to use a uniform schema for these objects. So, because `:source` is part of that schema, even though we ignore it for this type of module, we should allow a value to be passed for it.

mwaggett replied:
ah right, I forgot to circle back to this after reading the rest of the PR and understood things better.
yeah, I think it would be clearer to say something like
> Note: The `:source` parameter can be specified for Forge modules, but is currently ignored.

mwaggett commented on lib/r10k/environment/base.rb:
```diff
@@ -43,7 +43,7 @@ def initialize(name, basedir, dirname, options = {})
     @dirname = dirname
     @options = options
-    @puppetfile_name = options[:puppetfile_name]
+    @puppetfile_name = @options.delete(:puppetfile_name)
```
Would it make more sense to do `options.delete` instead of deleting from the global `@options`? It probably doesn't really matter, but I guess I would expect `@options` to contain _all_ the options, while `options` is only used in the scope of this method, so who cares if we delete from it.

reidmv replied:
Ooh, yes, and drawing attention to this caught another problem, 'cause the derived classes all use `options` now instead of `@options` ... will add a fix for this.

mwaggett commented on lib/r10k/environment/git.rb:
```diff
@@ -38,8 +40,16 @@ class R10K::Environment::Git < R10K::Environment::WithModules
+      # Standard option interface
+      :version => :ref,
+      :source  => :remote,
+      :type    => ::R10K::Util::Setopts::Ignore,
```
To make sure I understand this (since I don't have a ton of r10k experience) - is `:type` just used to determine which of these subclasses to use, and that's why we're ignoring it here?

reidmv replied:
It used to be for some objects, and after this PR is across the board, yeah. The r10k user interface is hella inconsistent. The big objective of this commit is to at least have a uniform common denominator interface, even though we have to stay backwards compatible with the old inconsistent stuff.

mwaggett commented on lib/r10k/environment/git.rb:
```diff
@@ -38,8 +40,16 @@ class R10K::Environment::Git < R10K::Environment::WithModules
+      # Type-specific options
+      :ref     => :self,
+      :remote  => :self,
+    }, raise_on_unhandled: false)
```
Why don't we want to raise on unhandled options here?

reidmv replied:
Because previously, git-based environments would let you pass garbage options. So for backwards compatibility, we still let you pass garbage options, here, for this one object type. 🤦

mwaggett replied:
oof. Want to add a comment to that effect? Then ideally we could remove that if we ever release a new major version.

reidmv replied:
Will do

mwaggett commented on lib/r10k/module/forge.rb:
```diff
@@ -32,13 +32,24 @@ def self.valid_version?(expected_version)
-  def initialize(title, dirname, expected_version, environment=nil)
+  include R10K::Util::Setopts
+
+  def initialize(title, dirname, opts, environment=nil)
```
I don't love that the third arg can now be two different things, but I don't know if I have a better idea right now. and this repo is probably full of that kind of thing, huh..

reidmv replied:
Yuuuup.

I also don't love that what _kind_ of arguments you gave `#mod` in a Puppetfile or a data hash, previously, was actually the main thing used to determine what _type_ of object you were specifying. Yay! But we need to keep it for backwards compatibility.

mwaggett replied:
Yay! 🙃

mwaggett commented on lib/r10k/module/svn.rb:
```diff
@@ -36,18 +36,21 @@ def self.implement?(name, args)
-    setopts(opts, INITIALIZE_OPTS)
+    setopts(opts, {
+      # Standard option interface
+      :source   => :url,
```
gah why do we use `:url` here and `:remote` in the environment version 😭 
(I know it was already like that, but still 😭)

mwaggett commented on spec/unit/module_spec.rb:
```diff
@@ -3,20 +3,36 @@
+    [
+      ['foo', {git: 'git url'}],
+      ['foo', {type: 'git', source: 'git url'}],
+    ].each do |scenario|
```
If you made this an array of maps, you could refer to the name and args separately, instead of using array indices.

reidmv replied:
True. Adding a minor re-work of these tests.

mwaggett commented on spec/unit/module_spec.rb:
```diff
@@ -3,20 +3,36 @@
+        expect(obj.send(:instance_variable_get, :'@expected_version')).to eq('8.0.0')
       end
     end
   end
```
Should there be svn tests? I notice there aren't any in this file and you didn't add any in `spec/unit/module/svn_spec.rb`.

reidmv replied:
May as well. Adding.

mwaggett commented on spec/unit/module/forge_spec.rb:
```diff
@@ -23,6 +23,12 @@
   end
 
+  describe "implementing the generic hash spec" do
+    it "should implement {type: forge}" do
```
Should this test include all the generic hash keys?

reidmv replied:
Might as well. Adding them.

mwaggett replied:
Not sure if you wanna update this description also

mwaggett commented on spec/unit/module/forge_spec.rb:
```diff
@@ -23,6 +23,12 @@
     end
   end
 
+  describe "implementing the generic hash spec" do
```
Do we need to test this for other module types too?

reidmv replied:
Not sure. I kind of trust `#setopts` to do its thing, at which point tests mostly sanity check that the `#setopts` definition of options is correct. A meaningful test would basically be a copy of that logic.

mwaggett replied:
That's fair. Would it make sense for this to go in the `base_spec` then? Or would that not work?

reidmv replied:
I think I put this test here specifically because the forge type was slightly more complicated than just calling setopts. The module_spec.rb already does basic initialization validation for several types.

mwaggett replied:
ah okay!

mwaggett commented on doc/puppetfile.mkd:
```diff
@@ -109,11 +109,19 @@ latest version available.
 
+An explicit type and/or version can be specified using the standard interface,
+`:type` and `:version`. The `:source` parameter is not supported for individual
+forge modules.
```
```suggestion
forge modules and will be ignored.
```

mwaggett commented on lib/r10k/environment/svn.rb:
```diff
@@ -44,8 +44,21 @@ class R10K::Environment::SVN < R10K::Environment::Base
+      :remote => :self,
+      :username => :self,
+      :password => :self,
+      :puppetfile_name => :self,
```
I take it this should be removed now, if it's being handled in the base class.

reidmv committed:
Add ignore, permissive capabilities to setopts

Let setopts explicity ignore some arguments, and give it the ability to
warn on options that don't match the schema, instead of raising. This is
to enable using setopts in places where we need to preserve backwards
compatibility with a less clearly defined, more permissive interface.

mwaggett reviewed (approved):
Not sure if you wanted to squash in the various pieces of that last improvements commit, but otherwise looks good!

@justinstoller did you want to take another look?
